feat(compute): let compute node handle deltas pushed from meta

soundOfDestiny opened the pull request:
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

feat(compute): let compute node handle deltas pushed from meta
this PR is successor of #4473 
in #4473  we let meta push latest version deltas to CN
here we let CN handle them, so CN will no longer need call `pin_version`

## Checklist

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* Installation and deployment 
* Connector (sources & sinks)
* SQL commands, functions, and operators
* RisingWave cluster configuration changes
* Other (please specify in the release note below)

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)

hzxa21 reviewed (commented):
Can you explain the background of this PR in the PR description?

soundOfDestiny commented:
> Can you explain the background of this PR in the PR description?

just edited

zwang28 commented on src/compute/src/server.rs:
```diff
@@ -141,6 +130,27 @@ pub async fn compute_node_serve(
+
+    let compute_observer_node =
+        ComputeObserverNode::new(filter_key_extractor_manager.clone(), local_version_manager);
+    // todo use ObserverManager
```
useless todo?

soundOfDestiny replied:
> useless todo?

fixed

zwang28 commented on src/compute/src/compute_observer/observer_manager.rs:
```diff
@@ -83,9 +106,13 @@ impl ObserverNodeImpl for ComputeObserverNode {
-    pub fn new(filter_key_extractor_manager: FilterKeyExtractorManagerRef) -> Self {
+    pub fn new(
+        filter_key_extractor_manager: FilterKeyExtractorManagerRef,
+        local_version_manager: Weak<LocalVersionManager>,
```
I think Weak ref is unnecessary, because after ObserverManager::start is joined, the whole instance is gone?

soundOfDestiny replied:
> I think Weak ref is unnecessary(Arc is fine), because after ObserverManager::start is joined, the whole instance is gone?

just in case

soundOfDestiny replied:
> I think Weak ref is unnecessary(Arc is fine), because after ObserverManager::start is joined, the whole instance is gone?

fixed

zwang28 commented on src/storage/src/hummock/local_version_manager.rs:
```diff
@@ -697,6 +699,7 @@ impl LocalVersionManager {
         }
     }
 
+    #[allow(dead_code)]
```
Please just remove `start_pin_worker `.

soundOfDestiny replied:
> Please just remove `start_pin_worker `.

fixed

zwang28 commented on src/storage/src/hummock/local_version_manager.rs:
```diff
@@ -223,10 +223,10 @@ impl LocalVersionManager {
-            Arc::downgrade(&local_version_manager),
-            hummock_meta_client.clone(),
-        ));
+        // tokio::spawn(LocalVersionManager::start_pin_worker(
```
Please just delete these unused code.

soundOfDestiny replied:
> Please just delete these unused code.

fixed

zwang28 commented on src/storage/src/monitor/monitored_store.rs:
```diff
@@ -32,7 +32,7 @@ use crate::{define_state_store_associated_type, StateStore, StateStoreIter};
 #[derive(Clone)]
 pub struct MonitoredStateStore<S> {
-    inner: S,
+    pub inner: S,
```
We can avoid this `pub`. Refer to Line 292 of this file.

soundOfDestiny replied:
> We can avoid this `pub`. Refer to Line 292 of this file.

fixed

zwang28 commented:
TODO:
To entirely removing pin_version, we need to:

- Fix ss_bench, that is add an observer manager for it too, to receive version update.
- (Can be done in another PR)In LocalVersionManager::new, use a placeholder `pinned_version` rather than one got by `pin_version`, because it will be replace by notification snapshot anyway. Meanwhile we need to fix the min_pin_version_id, because there is no longer an explicit pin_version when CN starts.

zwang28 reviewed (approved):
LGTM.

soundOfDestiny commented:
benchmark of this PR:
![localhost_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s (1)](https://user-images.githubusercontent.com/44948473/186856607-f250d3b5-170b-4d52-b11c-ec0a1c9f80b9.png)

soundOfDestiny commented:
benchmark of this PR:
![localhost_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s](https://user-images.githubusercontent.com/44948473/186857230-3846acb2-fcc6-4856-b38c-88813b5efc7e.png)

soundOfDestiny commented:
create source person
    (id INTEGER, name VARCHAR, "email_address" VARCHAR, "credit_card" VARCHAR, city VARCHAR, state VARCHAR, "date_time" TIMESTAMP)
with (
  connector = 'nexmark',
  nexmark.table.type = 'Person',
  nexmark.split.num = '24',
    nexmark.min.event.gap.in.ns = '0'
) row format json;

create source auction (id INTEGER, "item_name" VARCHAR, description VARCHAR, "initial_bid" INTEGER, reserve INTEGER, "date_time" TIMESTAMP, expires TIMESTAMP, seller INTEGER, category INTEGER) 
with (
  connector = 'nexmark',
  nexmark.table.type = 'Auction',
  nexmark.split.num = '24',
  nexmark.min.event.gap.in.ns = '0'
) row format json;

create source bid (auction INTEGER, bidder INTEGER, price INTEGER, "date_time" TIMESTAMP)
with (
  connector = 'nexmark',
  nexmark.table.type = 'Bid',
  nexmark.split.num = '24',
  nexmark.min.event.gap.in.ns = '0'
) row format json;

CREATE MATERIALIZED VIEW nexmark_q7 AS
SELECT
      B.auction,
      B.price,
      B.bidder,
      B.date_time
    FROM
      bid B
    JOIN (
      SELECT
        MAX(price) AS maxprice,
        window_end as date_time
      FROM
        TUMBLE(bid, date_time, INTERVAL '10' SECOND)
      GROUP BY
        window_end
    ) B1 ON B.price = B1.maxprice
    WHERE
      B.date_time BETWEEN B1.date_time - INTERVAL '10' SECOND
      AND B1.date_time;

soundOfDestiny commented:
benchmark without push scheme:
![localhost_3001_d_Ecy3uV1nz_risingwave_dashboard_orgId=1 refresh=10s (1)](https://user-images.githubusercontent.com/44948473/186871474-78a99784-6877-403d-a8b2-f41070b55dcf.png)
